### PR TITLE
AcceptMultipleTransactions: Fix workspace not being set as client_maxfeerate failure

### PR DESCRIPTION
**Pull request by instagibbs**
Bug causes an `Assume()` failure due to the expectation that the individual result should be invalid when done over `submitpackage` via rpc.

Bug introduced by https://github.com/bitcoin/bitcoin/pull/28950 , and I discovered it rebasing https://github.com/bitcoin/bitcoin/pull/28984 since it's easier to hit in that test scenario.

Tests in place were only checking `AcceptSingleTransaction`-level checks due to package evaluation only triggering when minfee is too high for the parent transaction.

Added test along with fix, moving the fill_mempool utility into a common area for re-use.

### CODE REVIEW

File: test/functional/rpc_packages.py [hunk 1]
```diff
@@ -369,23 +375,58 @@ def test_maxfeerate_maxburn_submitpackage(self):
...
+        minrate_btc_kvb_burn = min([chained_txn_burn["fee"] / chained_txn_burn["tx"].get_vsize() * 1000 for chained_txn_burn in chained_txns_burn])
+
         # Relax the restrictions for both and send it; parent gets through as own subpackage
-        pkg_result = node.submitpackage(chain_hex, maxfeerate=minrate_btc_kvb, maxburnamount=chained_txns[1]["new_utxo"]["value"])
-        assert "error" not in pkg_result["tx-results"][chained_txns[0]["wtxid"]]
+        pkg_result = node.submitpackage(chained_burn_hex, maxfeerate=minrate_btc_kvb_burn, maxburnamount=chained_txns_burn[1]["new_utxo"]["value"])
+        assert "error" not in pkg_result["tx-results"][chained_txns_burn[0]["wtxid"]]
         assert_equal(pkg_result["tx-results"][tx.getwtxid()]["error"], "scriptpubkey")
-        assert_equal(node.getrawmempool(), [chained_txns[0]["txid"]])
+        assert_equal(node.getrawmempool(), [chained_txns_burn[0]["txid"]])
```
Comment (glozow) [inline test/functional/rpc_packages.py:429]: unrelated renames?

File: src/validation.cpp [hunk 2]
```diff
@@ -1366,6 +1366,8 @@ PackageMempoolAcceptResult MemPoolAccept::AcceptMultipleTransactions(const std::
         // Individual modified feerate exceeded caller-defined max; abort
         // N.B. this doesn't take into account CPFPs. Chunk-aware validation may be more robust.
         if (args.m_client_maxfeerate && CFeeRate(ws.m_modified_fees, ws.m_vsize) > args.m_client_maxfeerate.value()) {
+            // Need to set failure here both individually and at package level
+            ws.m_state.Invalid(TxValidationResult::TX_MEMPOOL_POLICY, "max feerate exceeded", "");
             package_state.Invalid(PackageValidationResult::PCKG_TX, "max feerate exceeded");
```
Comment (glozow) [inline src/validation.cpp:1371]: Should this be "transaction failed" since it's an individual error?

**Review by glozow — COMMENTED** [review]
Concept ACK, verified that on master the test fails since the individual result wasn't populated and is "Valid"

File: test/functional/test_framework/util.py [hunk 3]
```diff
@@ -496,6 +496,48 @@ def check_node_connections(*, node, num_in, num_out):
     assert_equal(info["connections_in"], num_in)
     assert_equal(info["connections_out"], num_out)
 
+# Allows simpler testing of scenarios with minfee != minrelay
+def fill_mempool(test_framework, node, miniwallet):
```
[the remark this answers is not in the document]
Comment (instagibbs) [reply test/functional/test_framework/util.py:500]: it has `mine_large_block` et al which seem quite similar imo. I'll move it if there's something obviously better?

File: src/validation.cpp (hunk 2 above)
Comment (instagibbs) [reply src/validation.cpp]: changed

File: test/functional/test_framework/util.py [hunk 4]
```diff
@@ -496,6 +496,48 @@ def check_node_connections(*, node, num_in, num_out):
...
+        """Fill mempool until eviction."""
```
[the remark this answers is not in the document]
Comment (instagibbs) [reply test/functional/test_framework/util.py:504]: added some comments

**Comment by instagibbs** [comment]
Added some fuzz coverage which would have likely caught this

**Review by glozow — COMMENTED** [review]
Sorry for the nits, still need to run the fuzzer but otherwise lgtm

File: test/functional/rpc_packages.py [hunk 5]
```diff
@@ -31,6 +32,11 @@ def set_test_params(self):
         self.setup_clean_chain = True
         # whitelist peers to speed up tx relay / mempool sync
         self.noban_tx_relay = True
+        # Needed for fill_mempool
+        self.extra_args = [[
+            "-datacarriersize=100000",
+            "-maxmempool=5",
+        ]]
```
[the remark this answers is not in the document]
Comment (instagibbs) [reply test/functional/rpc_packages.py]: done

File: test/functional/test_framework/util.py (hunk 4 above)
[the remark this answers is not in the document]
Comment (instagibbs) [reply test/functional/test_framework/util.py]: oops yeah, done

**Review by glozow — COMMENTED** [review]
ACK 14c86ba721e1a208c88ada133ba9e90e24724ea4

Verified the fuzz test catches this problem, seems to run fine with the changes. Thanks for accepting the suggestions.

File: test/functional/test_framework/util.py [hunk 6]
```diff
@@ -496,6 +496,58 @@ def check_node_connections(*, node, num_in, num_out):
...
+        num_of_batches = 75
+        # Generate UTXOs to flood the mempool
+        # 1 to create a tx initially that will be evicted from the mempool later
+        # 75 transactions each with a fee rate higher than the previous one
+        # And 1 more to verify that this tx does not get added to the mempool with a fee rate less than the mempoolminfee
+        # And 2 more for the package cpfp test
+        test_framework.generate(miniwallet, 1 + (num_of_batches * tx_batch_size))
+
+        # Mine 99 blocks so that the UTXOs are allowed to be spent
+        test_framework.generate(node, 99)
```
Comment (murchandamus) [inline test/functional/test_framework/util.py:526]: In "Move fill_mempool to util function" (cc770e4b36e6f04cbd588ad3e7a19f122c21d38d):

Nit: If a commit is described as a move, I would expect no changes in the code beyond what’s necessary to update the callsites. While these changes are minor and obviously do not entail a functional change, I would have expected the updated commentary and refactor to be a separate commit—I’m wondering whether my understanding of the modus operandi is correct.

File: test/functional/rpc_packages.py [hunk 7]
```diff
@@ -369,23 +370,65 @@ def test_maxfeerate_maxburn_submitpackage(self):
...
+        pkg_result = node.submitpackage([parent["hex"], child["hex"]], maxfeerate=DEFAULT_FEE - Decimal("0.00000001"))
+
+        # Child is connected even though parent is invalid and still reports fee exceeded
+        # this implies sub-package evaluation of both entries together.
+        assert_equal(pkg_result["package_msg"], "transaction failed")
+        assert "mempool min fee not met" in pkg_result["tx-results"][parent["wtxid"]]["error"]
+        assert_equal(pkg_result["tx-results"][child["wtxid"]]["error"], "max feerate exceeded")
+        assert_equal(node.getrawmempool(), [])
+
+        # Reset minfee
```
Comment (murchandamus) [inline test/functional/rpc_packages.py:411]: In "AcceptMultipleTransactions: Fix workspace not being set as client_max…

…feerate failure" (https://github.com/bitcoin/bitcoin/commit/271599af6d84c871154dc16c7abcadbe5ac08163):

Did you perhaps mean:

```suggestion
        # Reset maxmempool and datacarriersize and empty mempool to reset dynamic mempool minimum feerate
```
?

**Review by murchandamus — COMMENTED** [review]
ACK 14c86ba721e1a208c88ada133ba9e90e24724ea4 with nits.

File: test/functional/rpc_packages.py [hunk 8]
```diff
@@ -369,23 +370,65 @@ def test_maxfeerate_maxburn_submitpackage(self):
...
+            fee_rate=DEFAULT_FEE,
+            utxo_to_spend=parent["new_utxo"],
+        )
+
+        pkg_result = node.submitpackage([parent["hex"], child["hex"]], maxfeerate=DEFAULT_FEE - Decimal("0.00000001"))
+
+        # Child is connected even though parent is invalid and still reports fee exceeded
+        # this implies sub-package evaluation of both entries together.
+        assert_equal(pkg_result["package_msg"], "transaction failed")
+        assert "mempool min fee not met" in pkg_result["tx-results"][parent["wtxid"]]["error"]
```
[the remark this answers is not in the document]
Comment (instagibbs) [reply test/functional/rpc_packages.py:407]: that's the error string as it's reported unfortunately!

File: test/functional/rpc_packages.py [hunk 9]
```diff
@@ -369,23 +370,65 @@ def test_maxfeerate_maxburn_submitpackage(self):
...
+        assert_equal(pkg_result["tx-results"][child["wtxid"]]["error"], "max feerate exceeded")
+        assert_equal(node.getrawmempool(), [])
+
+        # Reset minfee
+        self.restart_node(0)
+
```
[the remark this answers is not in the document]
Comment (instagibbs) [reply test/functional/rpc_packages.py:413]: done

File: test/functional/rpc_packages.py [hunk 10]
```diff
@@ -369,23 +370,65 @@ def test_maxfeerate_maxburn_submitpackage(self):
...
         assert_equal(pkg_result["tx-results"][chained_txns[1]["wtxid"]]["error"], "bad-txns-inputs-missingorspent")
         assert_equal(node.getrawmempool(), [])
 
+        # Make chain of two transactions where parent doesn't make minfee threshold
+        # but child is too high fee
+        # Needed for fill_mempool
+        self.restart_node(0, extra_args=[
+            "-datacarriersize=100000",
+            "-maxmempool=5",
+        ])
```
[the remark this answers is not in the document]
Comment (instagibbs) [reply test/functional/rpc_packages.py]: I don't want to blow away extra_args the caller may want, so took the comment suggestion for now

File: test/functional/test_framework/util.py (hunk 6 above)
Comment (instagibbs) [reply test/functional/test_framework/util.py:523]: split out the moving and docstring changes, should be minimal now?

File: test/functional/test_framework/util.py [hunk 11]
```diff
@@ -497,7 +497,15 @@ def check_node_connections(*, node, num_in, num_out):
     assert_equal(info["connections_out"], num_out)
 
 def fill_mempool(test_framework, node, miniwallet):
-        """Fill mempool until eviction."""
+        """Fill mempool until eviction.
+
+        Allows for simpler testing of scenarios with minfee != minrelay
+        Requires -datacarriersize=100000 and
+        -maxmempool=5(or similarly small).
```
Comment (glozow) [inline test/functional/test_framework/util.py:504]: "or similarly small" doesn't really make sense to me. AFAICT this makes just enough transactions to go past the 5MB mempool.

File: test/functional/test_framework/util.py [hunk 12]
```diff
@@ -497,7 +497,15 @@ def check_node_connections(*, node, num_in, num_out):
     assert_equal(info["connections_out"], num_out)
 
 def fill_mempool(test_framework, node, miniwallet):
-        """Fill mempool until eviction."""
+        """Fill mempool until eviction.
+
+        Allows for simpler testing of scenarios with minfee != minrelay
```
Comment (glozow) [inline test/functional/test_framework/util.py:502]: nit
```suggestion
        Allows for simpler testing of scenarios with minfee > minrelay
```

File: test/functional/rpc_packages.py [hunk 13]
```diff
@@ -411,6 +412,12 @@ def test_maxfeerate_maxburn_submitpackage(self):
         # Reset maxmempool and datacarriersize, and empty mempool to reset dynamic mempool minimum feerate
         self.restart_node(0)
 
+    def test_maxburn_submitpackage(self):
+        node = self.nodes[0]
+        # clear mempool
+        deterministic_address = node.get_deterministic_priv_key().address
+        self.generatetoaddress(node, 1, deterministic_address)
```
Comment (glozow) [inline test/functional/rpc_packages.py:419]: Mempool is already empty?

File: test/functional/rpc_packages.py [hunk 14]
```diff
@@ -369,23 +370,65 @@ def test_maxfeerate_maxburn_submitpackage(self):
...
+        # but child is too high fee
+        # Lower mempool limit to make it easier to fill_mempool
+        self.restart_node(0, extra_args=[
+            "-datacarriersize=100000",
+            "-maxmempool=5",
+        ])
+
+        fill_mempool(self, node, self.wallet)
+        while node.getrawmempool() != []:
+            self.generate(node, 1)
```
Comment (glozow) [inline test/functional/rpc_packages.py:391]: Note that this can change the mempool min fee

File: test/functional/rpc_packages.py [hunk 15]
```diff
@@ -369,23 +370,65 @@ def test_maxfeerate_maxburn_submitpackage(self):
...
+        pkg_result = node.submitpackage([parent["hex"], child["hex"]], maxfeerate=DEFAULT_FEE - Decimal("0.00000001"))
+
+        # Child is connected even though parent is invalid and still reports fee exceeded
+        # this implies sub-package evaluation of both entries together.
+        assert_equal(pkg_result["package_msg"], "transaction failed")
+        assert "mempool min fee not met" in pkg_result["tx-results"][parent["wtxid"]]["error"]
+        assert_equal(pkg_result["tx-results"][child["wtxid"]]["error"], "max feerate exceeded")
+        assert_equal(node.getrawmempool(), [])
+
+        # Reset maxmempool and datacarriersize, and empty mempool to reset dynamic mempool minimum feerate
```
Comment (glozow) [inline test/functional/rpc_packages.py:412]: Mempool is already empty
```suggestion
        # Reset maxmempool, datacarriersize, and dynamic mempool minimum feerate
```

File: test/functional/rpc_packages.py (hunk 14 above)
Comment (instagibbs) [reply test/functional/rpc_packages.py:391]: mm yeah let me make this a bit more bullet proof

Comment (glozow) [reply test/functional/rpc_packages.py:391]: Well I don't think you need get rid of the transactions here - you just want to make sure they get cleared by the end of the test right?

File: test/functional/test_framework/util.py (hunk 11 above)
Comment (instagibbs) [reply test/functional/test_framework/util.py]: removed the parenthesis

File: test/functional/test_framework/util.py (hunk 12 above)
Comment (instagibbs) [reply test/functional/test_framework/util.py]: taken

File: test/functional/rpc_packages.py (hunk 13 above)
Comment (instagibbs) [reply test/functional/rpc_packages.py]: replaced with an assert

File: test/functional/rpc_packages.py (hunk 14 above)
Comment (instagibbs) [reply test/functional/rpc_packages.py]: felt like I had a reason for this, but apparently not. At end of test I empty mempool on node restart, so removed

File: test/functional/rpc_packages.py (hunk 15 above)
Comment (instagibbs) [reply test/functional/rpc_packages.py]: changed where it's non-empty, so added `persistmempool=0` and left a new comment

File: test/functional/test_framework/util.py (hunk 6 above)
Comment (murchandamus) [reply test/functional/test_framework/util.py:523]: Nit: I don’t feel strongly about it, but the change that made me notice that there were code changes in the move originally was that  `COINBASE_MATURITY - 1` had gotten replaced with a `99` magic number.

File: src/validation.cpp [hunk 16]
```diff
@@ -1366,7 +1366,9 @@ PackageMempoolAcceptResult MemPoolAccept::AcceptMultipleTransactions(const std::
         // Individual modified feerate exceeded caller-defined max; abort
         // N.B. this doesn't take into account CPFPs. Chunk-aware validation may be more robust.
```
Comment (murchandamus) [inline src/validation.cpp:1367]: Nit: Subject line in commit message should not exceed 50 characters, and definitely not be more than 72: https://cbea.ms/git-commit/#limit-50

File: test/functional/test_framework/util.py [hunk 17]
```diff
@@ -497,7 +497,15 @@ def check_node_connections(*, node, num_in, num_out):
...
 def fill_mempool(test_framework, node, miniwallet):
-        """Fill mempool until eviction."""
+        """Fill mempool until eviction.
+
+        Allows for simpler testing of scenarios with floating minfee > minrelay
+        Requires -datacarriersize=100000 and
+        -maxmempool=5.
+        It will not ensure mempools become synced as it
+        is based on a single node and assumes -minrelaytxfee
+        is 1 sat/vbyte.
```
Comment (ismaelsadeeq) [inline test/functional/test_framework/util.py:507]: Maybe assert for this at the beginning of the `fill_mempool`
```python
     assert_equal(node.getnetworkinfo()['relayfee'], Decimal('0.00001000'))
```

**Review by murchandamus — COMMENTED** [review]
ACK 052c67d4beb72d3fdf31fa5e584ea2fa12e6f69c

Please feel free to ignore nits

**Review by ismaelsadeeq — COMMENTED** [review]
Concept ACK


I will run the new fuzz test with and without 72425bd699dd33d57794b3339f6266c4c9df443d

File: src/validation.cpp (hunk 16 above)
Comment (instagibbs) [reply src/validation.cpp:1367]: Made it shorter :)

File: test/functional/test_framework/util.py (hunk 6 above)
Comment (instagibbs) [reply test/functional/test_framework/util.py]: bit of an annoying circular reference issue(blocktools requiring utils) so for now at least make it grep-able what the magic value should be(and more greppable)

File: test/functional/test_framework/util.py [hunk 18]
```diff
@@ -497,7 +497,15 @@ def check_node_connections(*, node, num_in, num_out):
     assert_equal(info["connections_out"], num_out)
 
 def fill_mempool(test_framework, node, miniwallet):
-        """Fill mempool until eviction."""
+        """Fill mempool until eviction.
+
+        Allows for simpler testing of scenarios with floating minfee > minrelay
```
[the remark this answers is not in the document]
Comment (instagibbs) [reply test/functional/test_framework/util.py]: done

File: test/functional/test_framework/util.py [hunk 19]
```diff
@@ -496,6 +496,47 @@ def check_node_connections(*, node, num_in, num_out):
...
+        txouts = gen_return_txouts()
+        relayfee = node.getnetworkinfo()['relayfee']
+
+        tx_batch_size = 1
+        num_of_batches = 75
+        # Generate UTXOs to flood the mempool
+        # 1 to create a tx initially that will be evicted from the mempool later
+        # 75 transactions each with a fee rate higher than the previous one
+        # And 1 more to verify that this tx does not get added to the mempool with a fee rate less than the mempoolminfee
+        # And 2 more for the package cpfp test
```
[the remark this answers is not in the document]
Comment (instagibbs) [reply test/functional/test_framework/util.py]: done, I think

File: test/functional/test_framework/util.py (hunk 17 above)
Comment (instagibbs) [reply test/functional/test_framework/util.py:507]: done

**Review by ismaelsadeeq — APPROVED** [review]
ACK https://github.com/bitcoin/bitcoin/pull/29735/commits/4fe7d150eb3c85a6597d8fc910fe1490358197ad

I've reviewed [73b68....4fe7d15](https://github.com/bitcoin/bitcoin/compare/73b68bd8b4f9447e30091c7f8c3dc91a086bd93b..4fe7d150eb3c85a6597d8fc910f)

---

I've Ran the fuzz test without f28338b82be9820380ef217905ae9c167164f181

And verified the crash

```terminal
Test unit written to ./crash-7fd49c80d3088f672b4bf03ab8a1d7f1aedff0aa
Base64: q6sAIqsA/6sACQCrAAAAANr/+zUBAQENASUABAB5aKqgnQAAAAAlKv8tATUBAQENASUABAB5aKqgnQAAAAAluQAAAMP3AAAAAasAAAAAAAAA2v/7Kv8tATYBAQENASUABAB0aKKgle/f/wAA
```


I ran functional tests and `rpc_packages.py` failed.

---
I then cherry-picked the patch f28338b82be9820380ef217905ae9c167164f181 and run the fuzz test again, it passed successfully.
```terminal
INFO: seed corpus: files: 728 min: 1b max: 617355b total: 10276044b rss: 232Mb
#128    pulse  cov: 10579 ft: 39305 corp: 108/4319b exec/s: 64 rss: 260Mb
#512    pulse  cov: 11879 ft: 55511 corp: 356/67Kb exec/s: 36 rss: 283Mb
#925    INITED cov: 12041 ft: 66449 corp: 523/8522Kb exec/s: 20 rss: 550Mb
#925    DONE   cov: 12041 ft: 66449 corp: 523/8522Kb lim: 617355 exec/s: 20 rss: 550Mb
Done 925 runs in 46 second(s)
```

Functional tests are also passing

File: test/functional/test_framework/util.py (hunk 17 above)
Comment (ismaelsadeeq) [reply test/functional/test_framework/util.py:507]: looking at this again, what do you think about just restarting the node with the `-datacarriersize=100000`, ` -maxmempool=5`, `-minrelaytxfee=0.00001000` after filling the mempool we then restart the node with default node settings?
Are there any advantage of delegating setting this to the caller?
If not then this will reduce duplication of all callers of `fill_mempool` settings this values.

Comment (instagibbs) [reply test/functional/test_framework/util.py:507]: We can't restart the node after filling the memopool without resetting the floating minfee.

I could pull in the initial restart if we knew people didn't want any other extra_args.

Alternatively I could assert that datacarrier and maxmempool is set  in extra_args?

I think documentation is enough for now, leaving as is.

File: test/functional/rpc_packages.py [hunk 20]
```diff
@@ -369,23 +371,69 @@ def test_maxfeerate_maxburn_submitpackage(self):
...
+        # Child is connected even though parent is invalid and still reports fee exceeded
+        # this implies sub-package evaluation of both entries together.
+        assert_equal(pkg_result["package_msg"], "transaction failed")
+        assert "mempool min fee not met" in pkg_result["tx-results"][parent["wtxid"]]["error"]
+        assert_equal(pkg_result["tx-results"][child["wtxid"]]["error"], "max feerate exceeded")
+        assert parent["txid"] not in node.getrawmempool()
+        assert child["txid"] not in node.getrawmempool()
+
+        # Reset maxmempool, datacarriersize, ,reset dynamic mempool minimum feerate, and empty mempool.
+        self.restart_node(0, extra_args=["-persistmempool=0"])
```
Comment (glozow) [inline test/functional/rpc_packages.py:412]: This should be in the extra args _above_ to avoid writing mempool.dat.

Currently the test is fine, but this is because the new datacarriersize causes all the transactions to be rejected.

Comment (instagibbs) [reply test/functional/rpc_packages.py]: moved and added assertion of empty mempool after restart

**Review by glozow — COMMENTED** [review]
reACK 4ba1d0b55339c3ea90e2bcd64662a06f0f90dd46

**Review by theStack — APPROVED** [review]
ACK 4ba1d0b55339c3ea90e2bcd64662a06f0f90dd46

**Review by ismaelsadeeq — COMMENTED** [review]
re-ACK https://github.com/bitcoin/bitcoin/commit/4ba1d0b55339c3ea90e2bcd64662a06f0f90dd46  via [diff](https://github.com/bitcoin/bitcoin/compare/4fe7d150eb3c85a6597d8fc910fe1490358197ad..4ba1d0b55339c3ea90e2bcd64662a06f0f90dd46)
